lib/ast_ml.ml

```diff
@@ -1,134 +1,95 @@
-      }
-      | Match of {
-          data: {
-            matcher: t;
-            branches: (Pattern.t * t) list
-          },
-          position: Position.position
-        }
-      | Left of {
-        data of Match.t,
-        position: Position.position
+module rec S_P_TREE : sig
+  open Local_ast.L_P_TREE
+
+  type location = string
+
+  type tycon = string
+
+  and position = {
+    pos_fname : string;
+    pos_lnum : int;
+    pos_bol : int;
+    pos_cnum : int;
+  }
+
+  and pattern =
+    | Left of pattern
+    | Right of pattern
+    | Pattern of (location * lpattern)
+    | Pair of (pattern * pattern)
+    | Var of var
+    | Wild
+
+  and sugaredSend = 
+    | SendWithBinder of {
+        sender: location;
+        receiver: location;
+        binder: var;
+        expression: local_expr;
+        then_: choreography;
       }
-      | Right of {
-        data of Match.t,
-        position: Position.position
+    | SendWithoutBinder of {
+        sender: location;
+        receiver: location;
+        choreo: choreography;
       }
-  end
 
-  module Assignment : sig
-    type t =
-      | CType of {var: Var.t; tycon: Tycon.t}
-      | VarAsgn of {(Var.t * Tyvar.t option); location: Location.t option; choreo: Choreography.t}
-      | Fun of {var: Var.t; args: Pattern.t list; choreo: Choreography.t}
-  end
+  and syncData = {
+    sender: location;
+    receiver: location;
+    label: var;
+    then_: choreography;
+  }
+
+  and letData = {
+    declBlock: decl_block;
+    in_: choreography;
+  }
+
+  and matchData = {
+    matcher: choreography;
+    branches: (pattern * choreography) list;
+  }
+
+  and pair = (choreography * choreography)
+
+  and choreography =
+    | Unit
+    | ChoreoVar of {data: var; position: position}
+    | Expression of {data: (location * local_expr); position: position}
+    | Send of {data: sugaredSend; position: position}
+    | Branch of {data: (choreography * choreography * choreography); position: position}
+    | Sync of {data: syncData; position: position;}
+    | Let of {data: letData; position: position;}
+    | Pair of {data: pair; position: position}
+    | Fst of {data: pair; position: position}
+    | Snd of {data: pair; position: position}
+    | Match of {data: matchData; position: position;}
+    | Left of {data: matchData; position: position}
+    | Right of {data: matchData; position: position}
+
+  and assignment =
+    | CType of {var: var; tycon: tycon}
+    | VarAsgn of {var: (var * tycon option); location: location option; choreo: choreography}
+    | Fun of {var: var; args: pattern list; choreo: choreography}
 
-  module Dec : sig
-    type t =
-      | Fun of {var: Var.t; tyvars: Tyvar.t list}
-      | VarAsgn of {var: Var.t; tyvar: Tyvar.t; location: Location.t option}
-  end
+  and dec =
+    | FunDec of {var: var; tycons: tycon list}
+    | VarAsgnDec of {var: var; tycon: tycon; location: location option}
 
-  sharing type Exp.dec = Dec.t
+  and exp_dec = dec
```

Comment:
What's this doing?

Reply:
It never seems to be used anywhere, and I'm not sure what it's supposed to represent.